### PR TITLE
 mgr/dashboard: RGW multisite sync pipe

### PR DESCRIPTION
**Pull request by nmunet**
Fixes: https://tracker.ceph.com/issues/66926

Signed-off-by: Naman Munet <nmunet@redhat.com>

https://github.com/user-attachments/assets/674f405c-0041-4d86-80f7-a17b00af7cd9

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by cloudbehl** [comment]
Quick comment: can you add some info about flows and pipe. As a normal user I’m not able to understand what they mean and do.

**Comment by nmunet** [comment]
jenkins test dashboard cephadm

**Comment by nmunet** [comment]
jenkins test dashboard

**Comment by nmunet** [comment]
Screenshot 2024-07-16 at 12 05 05 AM

**Comment by nmunet** [comment]
jenkins test api

**Comment by nmunet** [comment]
jenkins test dashboard cephadm

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.html [hunk 1]
```diff
@@ -70,14 +70,48 @@
...
+          i18n>
+          Due to some internal dependencies these actions are disabled, it will get enabled once the issue gets resolved.
         </cd-alert-panel>
       </ng-template>
     </ng-container>
+    <ng-container ngbNavItem="pipe">
+      <a ngbNavLink
+         i18n>Pipe</a>
+      <ng-template ngbNavContent>
+        <legend>
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.html:86]:
```suggestion
        <legend i18n>
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.ts [hunk 2]
```diff
@@ -225,4 +280,67 @@ export class RgwMultisiteSyncPolicyDetailsComponent implements OnChanges {
...
+      size: 'lg'
+    });
+
+    try {
+      const res = await this.modalRef.result;
+      if (res === 'success') {
+        this.loadData();
+      }
+    } catch (err) {}
+  }
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.ts:302]: I don't really get what is this for

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 3]
```diff
@@ -2070,6 +2070,7 @@ def remove_sync_policy_group(self, group_id: str, bucket_name=''):
         except SubprocessError as error:
             raise DashboardException(error, http_status_code=500, component='rgw')
 
+    # pylint: disable=W0102
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/services/rgw_client.py:2073]: Why is this needed here?

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 4]
```diff
@@ -2150,7 +2156,7 @@ def create_sync_pipe(self, group_id: str, pipe_id: str,
     def remove_sync_pipe(self, group_id: str, pipe_id: str,
                          source_zones: Optional[List[str]] = None,
                          destination_zones: Optional[List[str]] = None,
-                         destination_buckets: Optional[List[str]] = None, bucket_name: str = ''):
+                         destination_bucket: Optional[List[str]] = None, bucket_name: str = ''):
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/services/rgw_client.py:2159]: The change is it to restrict it to a string rather than a list, right?
Same as you've done on the `create_sync_pipe`

```suggestion
                         destination_bucket: str = '', bucket_name: str = ''):
```

**Review by Pegonzal — COMMENTED** [review]
Thanks @nmunet for the PR!

I'm not entirely sure of the scope of this PR, but went over the code and left you some comments:

File: src/pybind/mgr/dashboard/frontend/cypress/e2e/page-helper.po.ts [hunk 5]
```diff
@@ -308,31 +308,31 @@ export abstract class PageHelper {
...
   }
 
-  searchNestedTable(tableSelector: string, text: string) {
+  searchNestedTable(selector: string, text: string) {
     this.waitDataTableToLoad();
 
     this.setPageSize('10');
-    cy.get(`${tableSelector} [aria-label=search]`).first().clear({ force: true }).type(text);
+    cy.get(`${selector} [aria-label=search]`).first().clear({ force: true }).type(text);
   }
```
[the remark this answers is not in the document]
Comment (nmunet) [reply src/pybind/mgr/dashboard/frontend/cypress/e2e/page-helper.po.ts:337]: In my previous PR I introduced these functions, and I thought it would be better if it can be generic and not specific, so did that!

File: src/pybind/mgr/dashboard/services/rgw_client.py (hunk 4 above)
Comment (nmunet) [reply src/pybind/mgr/dashboard/services/rgw_client.py:2159]: yes, maybe this got overwritten in the conflict resolution,
will fix it, thanks

File: src/pybind/mgr/dashboard/services/rgw_client.py (hunk 3 above)
Comment (nmunet) [reply src/pybind/mgr/dashboard/services/rgw_client.py:2073]: not sure, will remove

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.ts (hunk 2 above)
Comment (nmunet) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.ts:302]: once the modal is closed then I have to reload my datatable, its for that

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-pipe-modal/rgw-multisite-sync-pipe-modal.component.html [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  <ng-container
+    i18n="form title"
+    class="modal-title">{{ action | titlecase }} Pipe</ng-container>
+
+  <ng-container class="modal-content">
+    <form
+      name="pipeForm"
+      #frm="ngForm"
+      [formGroup]="pipeForm"
+      novalidate>
```
[the remark this answers is not in the document]
Comment (nmunet) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-pipe-modal/rgw-multisite-sync-pipe-modal.component.html:11]: will take care from now on!

**Review by cloudbehl — APPROVED** [review]
LGTM

**Comment by Pegonzal** [comment]
Looks like the Pipe is not editable I change the source zone and it doesn't update, the pipe name shows as editable but it is not so it should be readonly I suppose, also ideally, the notification's message would say 'updated' instead of 'create' :

[screen-capture (6).webm](https://github.com/user-attachments/assets/85fc0f5f-7458-4625-862d-92e4f7b0e9ad)

Kinda same thing in the Flows.
Personally would show the cd-alert-panel with the info message on top of the table, so it is next to the edit action that refers to.
The directional flow table doesn't show the name column, not sure if intended
[screen-capture (5).webm](https://github.com/user-attachments/assets/1eb99ec1-b20c-4aeb-b733-e72f5c63d482)

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-pipe-modal/rgw-multisite-sync-pipe-modal.component.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-pipe-modal/rgw-multisite-sync-pipe-modal.component.html:11]: Okay, marked unresolved since well... it is not resolved yet.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.ts (hunk 2 above)
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.ts:302]: makes sense, thanks

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-pipe-modal/rgw-multisite-sync-pipe-modal.component.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (nmunet) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-pipe-modal/rgw-multisite-sync-pipe-modal.component.html:11]: I thought it will have unnecessary changes in the PRs, so it would be better to keep the current style

**Comment by nmunet** [comment]
Thanks @Pegonzal ,
yes the pipe/flow update is only working in case we are adding zones instead of removing them, I have created a bug for it and handling it separately,
sure I will make other related changes that you suggested!

**Comment by nmunet** [comment]
jenkins test make check

**Comment by nmunet** [comment]
jenkins test dashboard cephadm

**Comment by nmunet** [comment]
jenkins test api

**Comment by nmunet** [comment]
jenkins test dashboard cephadm

**Comment by cloudbehl** [comment]
@nmunet make check is failing because of this:

`[lint:html    ] All files pass linting
[lint:prettier] src/app/ceph/rgw/rgw-multisite-sync-flow-modal/rgw-multisite-sync-flow-modal.component.ts
`

**Comment by nmunet** [comment]
jenkins test make check

**Comment by nmunet** [comment]
jenkins test dashboard

**Comment by nmunet** [comment]
jenkins test make check arm64

**Comment by nmunet** [comment]
jenkins test api

**Comment by nmunet** [comment]
jenkins test make check

**Comment by nmunet** [comment]
jenkins test make check

**Comment by nmunet** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-pipe-modal/rgw-multisite-sync-pipe-modal.component.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-pipe-modal/rgw-multisite-sync-pipe-modal.component.html:11]: I think there was a missunderstanding then. 

I wanted to say with `keep the current style` referring to use the same style we are using on the Dashboard.
With unnecessary file changes I meant on other PRs that will work with this code, since the code is not following current linting practices it will cause intances like this: https://github.com/ceph/ceph/pull/57715/files#diff-601d0819d8cbceeb03e674d6879d3c7b5566509783bf19b96abe356976271eb3
Where there are 200 line changes on the first file due to the linting differences. 

That's all. Hope it's more clear now what I meant

**Review by Pegonzal — APPROVED** [review]
PR looks goog,
there is still one thing from my side that was not addressed, the `Pipe Name` field would be better have it disabled when editing, rather than readonly, this way it is visible for the user that it is not for editing. But that we can take in other PR.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-pipe-modal/rgw-multisite-sync-pipe-modal.component.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (nmunet) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-pipe-modal/rgw-multisite-sync-pipe-modal.component.html:11]: oh, I misunderstood what you told before, do you want me to handle it in this PR, or can I do this in my other draft PR #58656 ?
